includes/multi-currency/Compatibility.php

```diff
@@ -319,6 +315,270 @@ public function convert_order_prices( $results ): array {
+	 *
+	 * @param array $updated_prices Prices updated by Product Add-Ons (unused).
+	 * @param array $cart_item      Cart item meta data.
+	 * @param array $prices         Original prices passed to Product Add-Ons for calculations.
+	 *
+	 * @return array
+	 */
+	public function update_product_price( $updated_prices, $cart_item, $prices ): array {
+		$price         = $this->multi_currency->get_price( $prices['price'], 'product' );
+		$regular_price = $this->multi_currency->get_price( $prices['regular_price'], 'product' );
+		$sale_price    = $this->multi_currency->get_price( $prices['sale_price'], 'product' );
+		$quantity      = $cart_item['quantity'];
+
+		// TODO: Check compat with Smart Coupons.
+		// Compatibility with Smart Coupons self declared gift amount purchase.
+		$credit_called = ! empty( $_POST['credit_called'] ) ? $_POST['credit_called'] : null;  // phpcs:ignore
+		if ( empty( $price ) && ! empty( $credit_called ) ) {
+			// Variable $_POST['credit_called'] is an array.
+			if ( isset( $credit_called[ $cart_item['data']->get_id() ] ) ) {
+				$price         = (float) $credit_called[ $cart_item['data']->get_id() ];
+				$regular_price = $price;
+				$sale_price    = $price;
+			}
+		}
+
+		if ( empty( $price ) && ! empty( $cart_item['credit_amount'] ) ) {
+			$price         = (float) $cart_item['credit_amount'];
+			$regular_price = $price;
+			$sale_price    = $price;
+		}
+
+		foreach ( $cart_item['addons'] as $addon ) {
+			// Percentage based and custom defined addon prices do not get converted, all others do.
+			if ( 'percentage_based' === $addon['price_type'] || 'custom_price' === $addon['field_type'] ) {
+				$addon_price = $addon['price'];
+			} elseif ( 'input_multiplier' === $addon['field_type'] ) {
+				// Quantity/multiplier add on needs to be split, calculated, then multiplied by input value.
+				$addon_price = $this->multi_currency->get_price( $addon['price'] / $addon['value'], 'product' ) * $addon['value'];
+			} else {
+				$addon_price = $this->multi_currency->get_price( $addon['price'], 'product' );
+			}
+
+			switch ( $addon['price_type'] ) {
+				case 'percentage_based':
+					$price         += (float) ( $cart_item['data']->get_price( 'view' ) * ( $addon_price / 100 ) );
+					$regular_price += (float) ( $regular_price * ( $addon_price / 100 ) );
+					$sale_price    += (float) ( $sale_price * ( $addon_price / 100 ) );
+					break;
+				case 'flat_fee':
+					$price         += (float) ( $addon_price / $quantity );
+					$regular_price += (float) ( $addon_price / $quantity );
+					$sale_price    += (float) ( $addon_price / $quantity );
+					break;
+				default:
+					$price         += (float) $addon_price;
+					$regular_price += (float) $addon_price;
+					$sale_price    += (float) $addon_price;
+					break;
+			}
+		}
+
+		// Let ourselves know this item has had add ons converted.
+		$cart_item['data']->update_meta_data( 'wcpay_mc_addons_converted', 1 );
```

Comment:
Minor nitpick: In Multi Currency Analytics we used the format `_wcpay_multi_currency_*` for order meta keys (e.g. `_wcpay_multi_currency_addons_converted`) - perhaps we could use the same here for consistency?

Also, it could be worth adding the meta key name as a constant at the top of the file and referencing the const here and wherever else we use this meta key. :)